cmd/query-tee/proxy.go

```diff
@@ -0,0 +1,73 @@
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+type Proxy struct {
+	cfg      Config
+	backends []*ProxyBackend
+	logger   log.Logger
+	metrics  *ProxyMetrics
+}
+
+func NewProxy(cfg Config, registerer prometheus.Registerer) (*Proxy, error) {
+	p := &Proxy{
+		cfg:     cfg,
+		logger:  util.Logger,
+		metrics: NewProxyMetrics(registerer),
+	}
+
+	// Parse the backend endpoints (comma separated).
+	parts := strings.Split(cfg.BackendEndpoints, ",")
+
+	for _, part := range parts {
+		u, err := url.Parse(part)
+		if err != nil {
+			return nil, errors.Wrapf(err, "invalid backend endpoint %s", part)
+		}
+
+		// The backend name is hardcoded as the backend hostname.
+		name := u.Hostname()
+
+		p.backends = append(p.backends, NewProxyBackend(name, u, cfg.BackendReadTimeout, name == cfg.PreferredBackend))
+	}
+
+	// At least 2 backends are required
+	if len(p.backends) < 2 {
+		return nil, errors.New("at least 2 backends are required")
+	}
+
+	return p, nil
+}
+
+func (p *Proxy) Run() error {
+	router := mux.NewRouter()
+
+	// Read endpoints.
+	router.Path("/api/prom/api/v1/query").Methods("GET").Handler(NewProxyEndpoint(p.backends, "api_v1_query", p.metrics, p.logger))
```

Comment:
This seems very Cortex-specific. Prometheus listens on `/api/v1/query`. Personally I think we should handle some `<prefix>/*`, where `/*` part would be appended to backend URLs, so that backend URL for Prometheus can be `http://prometheus/` and for Cortex it would be `http://cortex/api/prom/`. Query-tee datasource URL in Grafana would be `http://query-tee/prefix`. What do you think?

For route names and result comparison, I would recognize some well-known suffixes (like `/api/v1/query_range` and `/api/v1/label/[^/]+/values`) and ignore the rest.

Reply:
> Query-tee datasource URL in Grafana would be http://query-tee/prefix. What do you think?

I think your comment makes much sense. The `query-tee` now exposes Prometheus compatible APIs (ie. `/api/v1/query`). If the backend URL contains a path, that path is prepended to the original request. What's your take?

Reply:
> I think your comment makes much sense. The `query-tee` now exposes Prometheus compatible APIs (ie. `/api/v1/query`). If the backend URL contains a path, that path is prepended to the original request. What's your take?

This looks good. I'm little concerned about joining paths like that – I think it can lead to various anomalies [or not, haven't tested it] – but since this is a test tool, this will do.

Reply:
I've now used `path.Join()`. Better safe than sorry.

Reply:
> I've now used `path.Join()`. Better safe than sorry.

👍 Although I think it would still be possible to smuggle `../../../../../metrics` into request path :)